### PR TITLE
Add a new vector store hippo for langchain #11763

### PR DESCRIPTION
**Pull request by chocolate4**
Issue: #11763
Dependencies: N/A

I have run 'make format' 'make lint' and 'make test'

### CODE REVIEW

File: libs/langchain/langchain/vectorstores/hippo.py [hunk 1]
```diff
@@ -0,0 +1,637 @@
...
+            index_params=index_params,
+            drop_old=drop_old,
+            **kwargs,
+        )
+        logger.debug(f"[from_texts] texts:{texts}")
+        logger.debug(f"[from_texts] metadatas:{metadatas}")
+        vector_db.add_texts(texts=texts, metadatas=metadatas)
+        return vector_db
+
+# if __name__ == '__main__':
```
Comment (leo-gan) [inline libs/langchain/langchain/vectorstores/hippo.py:597]: Is it possible to add unit tests? This PR code is not simple, we need tests. 
After creating ut-s we don't need this commented code anymore.
Also, after creating ut-s, the `Chip War.txt` file should be moved to test folder. Now, we don't keep test data in the production code folders.

File: docs/docs_skeleton/docs/integrations/vectorstores/hippo.ipynb [hunk 2]
```diff
@@ -0,0 +1,459 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "source": [
+    "## Hippo\n",
+    "\n",
+    ">[Hippo](https://xxx) is a vector database with broad functionality.\n",
```
Comment (leo-gan) [inline docs/docs_skeleton/docs/integrations/vectorstores/hippo.ipynb:8]: Is it possible to add a real URL? `https://xxx` is confusing.

File: docs/docs_skeleton/docs/integrations/vectorstores/hippo.ipynb [hunk 3]
```diff
@@ -0,0 +1,459 @@
...
+    "query = \"Please introduce COVID-19\"\n",
+    "# query = \"Please introduce Hippo Core Architecture\"\n",
+    "# query = \"What operations does the Hippo Vector Database support for vector data?\"\n",
+    "# query = \"Does Hippo use hardware acceleration technology? Briefly introduce hardware acceleration technology.\"\n",
+    "\n",
+    "\n",
+    "# Retrieve similar content from the knowledge base,fetch the top two most similar texts.\n",
+    "res = vector_store.similarity_search(query, 2)\n",
+    "content_list = [item.page_content for item in res]\n",
+    "text = ''.join(content_list)"
```
Comment (leo-gan) [inline docs/docs_skeleton/docs/integrations/vectorstores/hippo.ipynb:340]: Is it possible to store the output of this code in the notebook (the `text` or `res`)? If it is big, then output only part of the text.

**Review by leo-gan — CHANGES_REQUESTED** [review]
Several nits with unit tests, etc.
This integration is great! Thank you @chocolate4 !!!

File: libs/langchain/langchain/vectorstores/hippo.py [hunk 4]
```diff
@@ -0,0 +1,637 @@
...
+            drop_old: Optional[bool] = False,
+    ):
+        self.number_of_shards = number_of_shards
+        self.number_of_replicas = number_of_replicas
+        self.embedding_func = embedding_function
+        self.table_name = table_name
+        self.database_name = database_name
+        self.index_params = index_params
+
+        # In order for a collection to be compatible, pk needs to be auto'id and int
```
[the remark this answers is not in the document]
Comment (TranswarpHippo) [reply libs/langchain/langchain/vectorstores/hippo.py:41]: I apologize for my oversight.I made a mistake in my comments here. Currently, Hippo only supports string type auto-increment primary keys. Also, in line 159 of the code, I have set the primary key to auto-increment when creating the table.

File: libs/langchain/langchain/vectorstores/hippo.py [hunk 5]
```diff
@@ -0,0 +1,637 @@
...
+                hosts = host.split(",")
+                given_address = ",".join([f"{h}:{port}" for h in hosts])
+            else:
+                given_address = str(host) + ":" + str(port)
+        else:
+            logger.debug("Missing standard address type for reuse atttempt")
+
+        try:
+            logger.info(f"create HippoClient[{given_address}]")
+            return HippoClient([given_address], username=username, pwd=password)
```
[the remark this answers is not in the document]
Comment (TranswarpHippo) [reply libs/langchain/langchain/vectorstores/hippo.py:92]: Yes, that's correct.

File: docs/docs_skeleton/docs/integrations/vectorstores/hippo.ipynb (hunk 3 above)
Comment (TranswarpHippo) [reply docs/docs_skeleton/docs/integrations/vectorstores/hippo.ipynb:340]: As I mentioned earlier, I have segmented the text into slices of 1000 tokens and only returned the top 2 contents. That is to say, the contents of both 'text' and 'res' will not exceed 2000 tokens, which is very small.Now, I will modify it to segment into slices of 500 tokens and return the top 2.

**Comment by TranswarpHippo** [comment]
this is my commit hash [08bfa1e](https://github.com/langchain-ai/langchain/pull/11840/commits/08bfa1e17e3e09cadd4d9ff2951b9d30d9e0283e).
The other commit hashes are those carried over from my fetch, rebase, and push operations.

**Review by leo-gan — APPROVED** [review]
lgtm

**Comment by chocolate4** [comment]
The last commit I fix spelling errors

**Comment by chocolate4** [comment]
The last commit I fix poetry.lock errors.

File: libs/langchain/langchain/vectorstores/hippo.py [hunk 6]
```diff
@@ -0,0 +1,693 @@
...
+            index_params=index_params,
+            drop_old=drop_old,
+            **kwargs,
+        )
+        logger.debug(f"[from_texts] texts:{texts}")
+        logger.debug(f"[from_texts] metadatas:{metadatas}")
+        vector_db.add_texts(texts=texts, metadatas=metadatas)
+        return vector_db
+
+# if __name__ == '__main__':
```
Comment (leo-gan) [inline libs/langchain/langchain/vectorstores/hippo.py:653]: Please, remove commented code.

**Comment by chocolate4** [comment]
@leo-gen Hello, is there a plan for when this section of code will be merged into the master branch?

**Comment by chocolate4** [comment]
@leo-gan  
Hello, is there a plan for when this section of code will be merged into the master branch?

File: libs/langchain/langchain/vectorstores/hippo.py [hunk 7]
```diff
@@ -0,0 +1,637 @@
+from transwarp_hippo_api.hippo_client import HippoTable, HippoClient, HippoField  # type: ignore
```
[the remark this answers is not in the document]
Comment (baskaryan) [reply libs/langchain/langchain/vectorstores/hippo.py:2]: these should all be conditionally imported within the functions that use them

**Comment by leo-gan** [comment]
@chocolate4 Please, fix the linting and we are ready to go.

**Comment by chocolate4** [comment]
hi @leo-gan I sincerely apologize for the inconvenience. I have resubmitted my code. Could you please review and merge it on your end? https://github.com/langchain-ai/langchain/pull/12412

**Pull request closed without merging**